src/fabtrimino.c

```diff
@@ -0,0 +1,250 @@
+#include "fabtrimino.h"
+
+struct square
+{
+    enum shape shape;
+};
+
+struct fabtrimino
+{
+    enum shape shape;
+    struct square*** matrix;
+    struct vector* pos;
+};
+
+
+void nullify_squares_matrix(struct square*** mat, size_t size)
+// TODO: catch and print error when size > matrix_size
+// (maybe create struct for matrix with size)
+{
+    for (int i = 0; i < size; i++)
+    {
+        for (int j = 0; j < size; j++)
+        {
+            mat[i][j] = NULL;
+        }
+    }
+}
+
+struct square* make_square(enum shape shape)
+{
+    struct square* sqr = calloc(1, sizeof(*sqr));
+    sqr->shape = shape;
+    return sqr;
+}
+
+void fill_T_squares(struct square*** matrix)
+{
+    matrix[0][2] = make_square(T);
+    matrix[1][2] = make_square(T);
+    matrix[2][2] = make_square(T);
+    matrix[1][1] = make_square(T);
+}
+
+void fill_I_squares(struct square*** matrix)
+{
+    matrix[0][1] = make_square(I);
+    matrix[1][1] = make_square(I);
+    matrix[2][1] = make_square(I);
+    matrix[3][1] = make_square(I);
+}
+
+void fill_J_squares(struct square*** matrix)
+{
+    matrix[0][1] = make_square(J);
+    matrix[0][2] = make_square(J);
+    matrix[1][2] = make_square(J);
+    matrix[2][2] = make_square(J);
+}
+
+void fill_L_squares(struct square*** matrix)
+{
+    matrix[0][2] = make_square(L);
+    matrix[1][2] = make_square(L);
+    matrix[2][2] = make_square(L);
+    matrix[2][1] = make_square(L);
+}
+
+void fill_O_squares(struct square*** matrix)
+{
+    matrix[1][1] = make_square(O);
+    matrix[1][2] = make_square(O);
+    matrix[2][1] = make_square(O);
+    matrix[2][2] = make_square(O);
+}
+
+void fill_S_squares(struct square*** matrix)
+{
+    matrix[1][1] = make_square(S);
+    matrix[2][1] = make_square(S);
+    matrix[0][2] = make_square(S);
+    matrix[1][2] = make_square(S);
+}
+
+void fill_Z_squares(struct square*** matrix)
+{
+    matrix[0][1] = make_square(Z);
+    matrix[1][1] = make_square(Z);
+    matrix[1][2] = make_square(Z);
+    matrix[2][2] = make_square(Z);
+}
+
+void fill_shape_squares(struct square*** matrix, enum shape shape)
+{
+    switch(shape){
+    case T:
+        fill_T_squares(matrix);
+        break;
+    case I:
+        fill_I_squares(matrix);
+        break;
+    case J:
+        fill_J_squares(matrix);
+        break;
+    case L:
+        fill_L_squares(matrix);
+        break;
+    case Z:
+        fill_Z_squares(matrix);
+        break;
+    case S:
+        fill_S_squares(matrix);
+        break;
+    case O:
+        fill_O_squares(matrix);
+        break;
+    default:
+        printf("INVALID SHAPE: %d\n", shape);
+        break;
+    }
+}
+
+struct fabtrimino* fab_make(enum shape shape)
+{
+    struct fabtrimino* fab = calloc(1, sizeof(*fab));
+
+    fab->matrix = calloc(4, sizeof(struct square**));
```

Comment:
shouldn't your matrix be 2x3 and not 4x4? Or I guess the 4x4 includes the rotations?

Reply:
It's difficult to squeeze the 1x4 bar in a 2x3 matrix.